iradix_test.go

```diff
@@ -212,6 +211,145 @@ func TestDelete(t *testing.T) {
 	}
 }
 
+func TestDeletePrefix(t *testing.T) {
+
+	type exp struct {
+		treeNodes   []string
+		prefix      string
+		expectedOut []string
+	}
+
+	//various test cases where DeletePrefix should succeed
+	cases := []exp{
+		{[]string{"", "test/test1", "test/test2", "test/test3", "R", "RA"}, "test", []string{"", "R", "RA"}},
+		{[]string{"", "test", "test/test1", "test/test2", "test/test3", "test/testAAA", "R", "RA"}, "test", []string{"", "R", "RA"}},
+		{[]string{"", "test/test1", "test/test2", "test/test3", "test/testAAA", "R", "RA"}, "test/test", []string{"", "R", "RA"}},
+		{[]string{"", "AB", "ABC", "AR", "R", "RA"}, "AR", []string{"", "AB", "ABC", "R", "RA"}},
+	}
+
+	for _, testCase := range cases {
+		verifyPrefixDelete(t, testCase.treeNodes, testCase.prefix, testCase.expectedOut)
+	}
+
+	//verify deleting a prefix that doesn't exist
+	verifyInvalidPrefix(t, []string{"", "test/test1", "test/test2"}, "XXXX")
+}
+func verifyPrefixDelete(t *testing.T, nodes []string, prefix string, expected []string) {
+	r := New()
+	for _, ss := range nodes {
+		r, _, _ = r.Insert([]byte(ss), true)
+	}
+	if r.Len() != len(nodes) {
+		t.Fatalf("Unexpected tree length after insert, expected %v but got %v ", len(nodes), r.Len())
+	}
+	r, ok := r.DeletePrefix([]byte(prefix))
+	if !ok {
+		t.Fatalf("DeletePrefix should have returned true for tree %v, deleting prefix %v", nodes, prefix)
+	}
+	if r.Len() != len(expected) {
+		t.Fatalf("Bad tree length, expected %v got %v for tree %v, deleting prefix %v ", len(expected), r.Len(), nodes, prefix)
+	}
+
+	verifyTree(t, expected, r)
+
+}
+
+func verifyInvalidPrefix(t *testing.T, nodes []string, prefix string) {
+	r := New()
+	for _, ss := range nodes {
+		r, _, _ = r.Insert([]byte(ss), true)
+	}
+
+	prevLen := r.Len()
+	r, ok := r.DeletePrefix([]byte(prefix))
+	if ok {
+		t.Fatal("DeletePrefix should have returned false")
+	}
+	if r.Len() != prevLen {
+		t.Fatalf("Bad tree length, expected %v but got %v", prevLen, r.Len())
+	}
+}
+
+func TestTrackMutate_DeletePrefix(t *testing.T) {
+
+	r := New()
+
+	keys := []string{
+		"foo",
+		"foo/bar/baz",
+		"foo/baz/bar",
+		"foo/zip/zap",
+		"bazbaz",
+		"zipzap",
+	}
+	for _, k := range keys {
+		r, _, _ = r.Insert([]byte(k), nil)
+	}
+	if r.Len() != len(keys) {
+		t.Fatalf("bad len: %v %v", r.Len(), len(keys))
+	}
+
+	rootWatch, _, _ := r.Root().GetWatch(nil)
+	if rootWatch == nil {
+		t.Fatalf("Should have returned a watch")
+	}
+
+	nodeWatch, _, _ := r.Root().GetWatch([]byte("foo/bar/baz"))
```

Comment:
For this it might be nice to get all three of the foo ones to make sure we recurse and get all of them.